### PR TITLE
Move from CircleCI to GitHub Actions for Continuous Integration

### PR DESCRIPTION
**Pull request by tunetheweb**
### Brief summary of the change made

This PR adds running of our CI tests in GitHub CI in place of CircleCI.

There are a number of reasons for moving away form CircleCI including:
- CicrlceCI recently blocked us for no good reason and it took 3 days to sort it out
- GitHub Action is free for public repos like ours
- We already have a dependency on GitHub and other GitHub Actions.
- What we use on sqlfluff-online
- Lots of integration options and big marketplace of pre-built actions you can use.
- Also us to bring back Windows runners for each commit
- Less concurrency limits meaning different PRs can run in parallel
- Fully integrated into GitHub so no launching off to a separate site.
- Doesn’t require giving access to private repos to see results (better privacy for our contributors)
- Allows us to only run the tests on code changes, and nor simple README changes for example.

### Are there any other side effects of this change that we should be aware of?

- [ ] - We need to remove branch protections for `main` branch which currently insist Circle CI tests complete as they will not run any more. I propose not adding the equivalent GiHub tests in so simple changes (e.g. images or ReadMe) do not need these tests run. Unfortunately GitHub does not have a "any tests which are run must complete successfully, but don't flag if a test is not run" option, which would be better than "these tests must always run even if not applicable" IMHO.
- [ ] - We also need to turn this off at Circle CI as currently CircleCI complains there is no CircleCI config file anymore.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/parser` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: .github/workflows/ci-tests.yml
Comment (barrywhart) [inline .github/workflows/ci-tests.yml:45]: Want to consider fewer versions here as you mentioned elsewhere?

**Review by barrywhart — APPROVED** [review]
Looks good! One small question.

File: .github/workflows/ci-tests.yml
Comment (tunetheweb) [reply .github/workflows/ci-tests.yml:45]: I think I'd like to keep it as is for now, in case there's any version specific things in there (either now or in future). I see 3.6 has different parallel processing for example.

So let's leave this as a bigger discussion. And maybe it's not an issue if we can get on CircleCI as an open source project.

**Comment by tunetheweb** [comment]
I can't merge this as currently the CircleCI jobs are mandatory for merging to `main` in branch protection settings. Any of the @sqlfluff/sqlfluff-maintainers able to (temporarily) turn those off?

**Comment by NiallRees** [comment]
> I can't merge this as currently the CircleCI jobs are mandatory for merging to `main` in branch protection settings. Any of the @sqlfluff/sqlfluff-maintainers able to (temporarily) turn those off?

I think this needs @alanmcruickshank

**Review by NiallRees — APPROVED** [review]
Interestingly, dbt just switched to GH actions from Circle: https://github.com/dbt-labs/dbt/issues/3688

File: .github/workflows/doclinting.yml
Comment (alanmcruickshank) [inline .github/workflows/doclinting.yml:7]: Typo

```suggestion
# We pull Doc Linting out separately so it can be run independently
```

File: README.md
Comment (alanmcruickshank) [inline README.md:13]: Did you mean this url instead?

```suggestion
[![CI Tests](https://github.com/sqlfluff/sqlfluff/workflows/CI%20Tests/badge.svg)](https://github.com/sqlfluff/sqlfluff/actions/workflows/ci-tests.yml)
```

File: tox.ini
Comment (alanmcruickshank) [inline tox.ini:38]: Typo

```suggestion
# is often on different drive (C) from working dir (D), which causes problems.

```

**Review by alanmcruickshank — APPROVED** [review]
I've disabled the branch protection rules for now, and I can re-enable them on the new checks once this is merged. Couple of nits, but I think the thrust of this is all good 👍. Thanks for taking the time to put this together.

File: README.md
Comment (tunetheweb) [reply README.md:13]: Oops copy and paste error form my other project that I stole this from. But I don't think it's quite right still though. I think this is what we need:

```suggestion
[![CI Tests](https://github.com/sqlfluff/sqlfluff/workflows/CI%20Tests/badge.svg)](https://github.com/sqlfluff/sqlfluff/actions?query=workflow%3A%22CI+Tests%22)
```

**Comment by tunetheweb** [comment]
> I've disabled the branch protection rules for now, and I can re-enable them on the new checks once this is merged.

No you can't, as it is at present.

The new workflow does not run for Markdown or Images-only changes. So the tests will not complete. So you won't be able to megre.

The CI Tests branch also does not run for `.rst` files (doclint has a separate workflow).

So our choices are:
1. Disable branch protections permanently. Up to Maintainers to decide if decide whether to merge even if all tests don't pass (maybe we're not comfortable with that?).
2. Merge doclint back into main CI workflow, and have it run on every commit, even if just a markdown or image change (probably not the worst to be honest given most of our changes are not code changes)
3. Have some fancy logic in the workflow to run, but skip most of the steps, for certain changes (this I think will be quite complicated)

This PR is currently option 1, originally to allow it to be switched back and forth with CircleCI, but maybe now we're ditching Circle CI and going all in on GitHub Action root we should go with option 2?

Will commit that.

**Comment by tunetheweb** [comment]
OK should be good to merge now @alanmcruickshank and re-enable branch protections on the new jobs once it's been merged.

Don't forget to turn off at the Circle Ci side too or you get a failed job each time.

**Comment by tunetheweb** [comment]
Merged - you can re-enable branch protections now @alanmcruickshank

**Pull request merged**
